packages/features/auth/signup/handlers/calcomHandler.ts

```diff
@@ -0,0 +1,193 @@
+
+import { joinOrganization, joinAnyChildTeamOnOrgInvite } from "../utils/organization";
+import { findTokenByToken, throwIfTokenExpired, validateUsernameForTeam } from "../utils/token";
+
+async function handler(req: RequestWithUsernameStatus, res: NextApiResponse) {
+  const {
+    email: _email,
+    password,
+    token,
+  } = signupSchema
+    .pick({
+      email: true,
+      password: true,
+      token: true,
+    })
+    .parse(req.body);
+  let username: string | null = req.usernameStatus.requestedUserName;
+  let checkoutSessionId: string | null = null;
+
+  // Check for premium username
+  if (req.usernameStatus.statusCode === 418) {
+    return res.status(req.usernameStatus.statusCode).json(req.usernameStatus.json);
+  }
+
+  // Validate the user
+  if (!username) {
+    throw new HttpError({
+      statusCode: 422,
+      message: "Invalid username",
+    });
+  }
+
+  const email = _email.toLowerCase();
+
+  let foundToken: { id: number; teamId: number | null; expires: Date } | null = null;
+  if (token) {
+    foundToken = await findTokenByToken({ token });
+    throwIfTokenExpired(foundToken?.expires);
+    validateUsernameForTeam({ username, email, teamId: foundToken?.teamId });
+  } else {
+    const usernameAndEmailValidation = await validateUsername(username, email);
+    if (!usernameAndEmailValidation.isValid) {
+      throw new HttpError({
+        statusCode: 409,
+        message: "Username or email is already taken",
+      });
+    }
+  }
+
+  // Create the customer in Stripe
+  const customer = await stripe.customers.create({
+    email,
+    metadata: {
+      email /* Stripe customer email can be changed, so we add this to keep track of which email was used to signup */,
+      username,
+    },
+  });
+
+  const returnUrl = `${WEBAPP_URL}/api/integrations/stripepayment/paymentCallback?checkoutSessionId={CHECKOUT_SESSION_ID}&callbackUrl=/auth/verify?sessionId={CHECKOUT_SESSION_ID}`;
+
+  // Pro username, must be purchased
+  if (req.usernameStatus.statusCode === 402) {
+    const checkoutSession = await stripe.checkout.sessions.create({
+      mode: "subscription",
+      payment_method_types: ["card"],
+      customer: customer.id,
+      line_items: [
+        {
+          price: getPremiumMonthlyPlanPriceId(),
+          quantity: 1,
+        },
+      ],
+      success_url: returnUrl,
+      cancel_url: returnUrl,
+      allow_promotion_codes: true,
+    });
+
+    /** We create a username-less user until he pays */
+    checkoutSessionId = checkoutSession.id;
+    username = null;
+  }
+
+  // Hash the password
+  const hashedPassword = await hashPassword(password);
+
+  if (foundToken && foundToken?.teamId) {
+    const team = await prisma.team.findUnique({
+      where: {
+        id: foundToken.teamId,
+      },
+    });
+    if (team) {
+      const teamMetadata = teamMetadataSchema.parse(team?.metadata);
+
+      const user = await prisma.user.upsert({
```

Comment:
This should go in a transaction since we are updating multiple records. If any fails, we'll have data in a bad state.